File: 14888.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;

int N;
int maxNum = INT_MIN;
int minNum = INT_MAX;
int nums[11];

void dfs(int numPlus, int numMinus, int numTimes, int numDiv, int cnt, int res) {
    if (cnt == N) {
        maxNum = max(maxNum, res);
        minNum = min(minNum, res);
        return;
    }
    if  (numPlus) dfs(numPlus - 1, numMinus, numTimes, numDiv, cnt + 1, res + nums[cnt]);
    if (numMinus) dfs(numPlus, numMinus - 1, numTimes, numDiv, cnt + 1, res - nums[cnt]);
    if (numTimes) dfs(numPlus, numMinus, numTimes - 1, numDiv, cnt + 1, res * nums[cnt]);
    if   (numDiv) dfs(numPlus, numMinus, numTimes, numDiv - 1, cnt + 1, res / nums[cnt]);
}

int main() {
    int temp;
    int numPlus, numMinus, numTimes, numDiv;
    
    cin >> N;
    
    for(int i = 0; i < N; i++) {
        cin >> nums[i];
    }
    
    cin >> numPlus >> numMinus >> numTimes >> numDiv;
    
    dfs(numPlus, numMinus, numTimes, numDiv, 1, nums[0]);
    
    cout << maxNum << endl << minNum;
}

File: 14888.py
import sys

input = lambda: sys.stdin.readline().strip()

n = int(input())
arr = list(map(int, input().split()))
num_plus, num_minus, num_multiply, num_division = map(int, input().split())

max_num = 0
min_num = float('inf')

def dfs(pos, num_plus, num_minus, num_multiply, num_division, res):
    global min_num, max_num
    if pos == n:
        min_num = min(min_num, res)
        max_num = max(max_num, res)
        return
    
    if num_plus:     dfs(pos + 1, num_plus - 1, num_minus, num_multiply, num_division, res + arr[pos])
    if num_minus:    dfs(pos + 1, num_plus, num_minus - 1, num_multiply, num_division, res - arr[pos])
    if num_multiply: dfs(pos + 1, num_plus, num_minus, num_multiply - 1, num_division, res * arr[pos])
    if num_division: dfs(pos + 1, num_plus, num_minus, num_multiply, num_division - 1, int(res / arr[pos]))
    
dfs(1, num_plus, num_minus, num_multiply, num_division, arr[0])
print(max_num)
print(min_num)